torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -68,244 +72,255 @@ class SmemAllocMap {
+// location prefered for the sync threads is the last possible position. One
+// future optimization could be to not sync on the last iteration of the loop
+// the sync is placed in.
+class WarSyncInserter : private kir::ExprMutator {
+ public:
+  static std::vector<kir::Expr*> insert(const std::vector<kir::Expr*>& exprs) {
+    WarSyncInserter inserter(exprs);
+    return inserter.exprs_;
   }
 
-  void handle(kir::Expr* expr) final {
-    if (ir_utils::isTVOp(expr)) {
-      is_last_op_sync_ = false;
-
-      // For this SyncInserter
-      if (initial_sync_) {
-        addInputSmemTvs(expr, final_);
-      } else {
-        addInputSmemTvs(expr, final_);
-        addOutputSmemTvs(expr, initial_);
-      }
-
-      // For parent SyncInserter
-      addOutputSmemTvs(expr, all_smem_outputs_);
-      addInputSmemTvs(expr, all_smem_inputs_);
-    } else {
-      kir::ExprMutator::handle(expr);
+ private:
+  //! Insert Sync nodes at the end of a given for-loop when a WAR
+  //! hazard may happen.
+  WarSyncInserter(const std::vector<kir::Expr*>& exprs) {
+    auto& lower_alloc_info_map = GpuLower::current()->localAllocationInfoMap();
+    for (const auto& entry : lower_alloc_info_map) {
+      alloc_map_.insert(entry.first);
     }
+    kir::ExprMutator::traverseAndInsert(exprs);
   }
 
-  void handle(kir::Allocate* alloc) final {
-    alloc_map_.insert(alloc);
+  void handle(kir::IfThenElse* ite) final {
+    TORCH_INTERNAL_ASSERT(
+        ite->elseBody().empty(),
+        "Pass does not support conditional flow,",
+        " needs to be done before conditional execution is lowered.");
+    kir::ExprMutator::handle(ite);
   }
 
   void handle(kir::Sync* sync) final {
-    is_last_op_sync_ = true;
-    initial_sync_ = true;
-    final_.clear();
+    // Register the sync for the active for loop
+    sync_hit_.back() = true;
+    // Run through the active allocations, if a read was hit, register there was
+    // a sync after the read. If there's subsequent reads on this buffer the
+    // sync_after_read will be cleared.
+    for (auto& entry : smem_allocations_) {
+      auto& alloc_stack = entry.second;
+      if (alloc_stack.back().read_hit) {
+        alloc_stack.back().sync_after_read = true;
+      }
+    }
   }
 
-  void handle(kir::ForLoop* fl) final {
-    if (fl == base_fl_) {
-      kir::ExprMutator::handle(fl);
-      return;
+  // Checks if fl or loops within it have hit a sync
+  bool syncWithin(kir::ForLoop* fl) {
+    // If outer most scope check the first sync_hit_ position
+    if (fl == nullptr) {
+      return sync_hit_[0];
     }
 
-    LocalSyncInserterForLoop child_sync_inserter(fl, alloc_map_);
-
-    const auto& child_inputs = child_sync_inserter.all_smem_inputs();
-    const auto& child_outputs = child_sync_inserter.all_smem_outputs();
-    const bool maybe_skipped = !fl->start()->isZeroInt() &&
-        !isParallelTypeThread(fl->iter_domain()->parallelType());
```

Comment:
Looks like this part of the previous logic is missing in the PR. This was to address loops like:

```
for (int i = 1; i < N; ++i) {
  ___barrier_sync(0);
  ...
}
```
Here, since `N` can be just `1`, it is possible that the loop body is not visited at all.

The reason it checks if the loop is parallelized is because when it's parallelized, `kir::Sync` expressions inside such loops should be always visited as we generate codes as follows:

```
___barrier_sync(0);
if (tid >= 1 && tid <N) {
  ...
}
```

I don't think of any actual case where this pattern could appear, but strictly speaking I think the previous conservative analysis is required.

Reply:
An easy fix may be to just change the codegen so that for-loops always start from 0 regardless of actual start values.

Reply:
I don't think this is a concern that is realistic enough to block this PR, so I'll approve the PR.

Reply:
Started all loops at 0.